tests/test_notify_v2.py

```diff
@@ -0,0 +1,415 @@
+        post = self.s3.generate_presigned_post(
+            Bucket=bucket,
+            Key=key,
+            ExpiresIn=60,
+            Fields={'Content-Type': "application/json"},
+            Conditions=[{'Content-Type': "application/json"}]
+        )
+        subscription = self._put_subscription(
+            {
+                'payload_form_field': "file",
+                'form_fields': post['fields'],
+                'callback_url': post['url'],
+                'encoding': "multipart/form-data",
+            },
+            replica
+        )
+
+        queue_notification(
+            replica,
+            subscription,
+            "CREATE",
+            "bundles/a47b90b2-0967-4fbf-87bc-c6c12db3fedf.2017-07-12T055120.037644Z",
+            delay_seconds=0
+        )
+        notification = self._get_notification_from_s3_object(bucket, key)
+        self.assertEquals(notification['subscription_id'], subscription['uuid'])
+
+    @testmode.integration
+    def test_bundle_notification(self):
+        for replica in Replica:
+            with self.subTest(replica):
+                self._test_bundle_notification(replica)
+
+    def _test_bundle_notification(self, replica):
+        bucket = get_env('DSS_S3_BUCKET_TEST')
+        key = f"notification-v2/{uuid4()}"
+        url = self.s3.generate_presigned_url(
+            ClientMethod='put_object',
+            Params=dict(Bucket=bucket, Key=key, ContentType="application/json")
+        )
+        subscription = self._put_subscription(
+            {
+                'callback_url': url,
+                'method': "PUT",
+            },
+            replica
+        )
+
+        # upload test bundle from test fixtures bucket
+        bundle_uuid, bundle_version = self._upload_bundle(replica)
+
+        notification = self._get_notification_from_s3_object(bucket, key)
+        self.assertEquals(notification['subscription_id'], subscription['uuid'])
+        self.assertEquals(notification['match']['bundle_uuid'], bundle_uuid)
+        self.assertEquals(notification['match']['bundle_version'], bundle_version)
+
+    @eventually(30, 1, {botocore.exceptions.ClientError})
+    def _get_notification_from_s3_object(self, bucket, key):
+        obj = self.s3.get_object(Bucket=bucket, Key=key)['Body'].read().decode("utf-8")
+        return json.loads(obj)
+
+    @testmode.standalone
+    def test_subscription_enumerate(self, replica=Replica.aws):
+        url = str(UrlBuilder()
+                  .set(path="/v1/subscriptions")
+                  .add_query("replica", replica.name)
+                  .add_query("subscription_type", "jmespath"))
+        self.assertGetResponse(
+            url,
+            requests.codes.ok,
+            headers=get_auth_header())
+
+    @testmode.integration
+    def test_get_subscriptions_for_replica(self):
+        for replica in Replica:
+            with self.subTest(replica.name):
+                subscriptions_v2.get_subscriptions_for_replica(replica)
```

Comment:
as the number of subscribers increases how will this scale or are we making the assumption that there will not be a lot of subscribers?

Reply:
Optimization may be required if the number of subscribers grows large. This is a tradeoff of avoiding Elasticsearch and percolate queries.